packages/typescript-estree/src/parser.ts

```diff
@@ -566,12 +572,43 @@ function parseAndGenerateServices<T extends TSESTreeOptions = TSESTreeOptions>(
    */
   const shouldProvideParserServices =
     extra.programs != null || (extra.projects && extra.projects.length > 0);
-  const { ast, program } = getProgramAndAST(
-    code,
-    extra.programs,
-    shouldProvideParserServices,
-    extra.createDefaultProgram,
-  )!;
+
+  /**
+   * If we are in singleRun mode but the parseAndGenerateServices() function has been called more than once for the current file,
+   * it must mean that we are in the middle of an ESLint automated fix cycle (in which parsing can be performed up to an additional
+   * 10 times in order to apply all possible fixes for the file).
+   *
+   * In this scenario we cannot rely upon the singleRun AOT compiled programs because the SourceFiles will not contain the source
+   * with the latest fixes applied. Therefore we fallback to creating the quickest possible isolated program from the updated source.
+   */
+  let ast: ts.SourceFile;
+  let program: ts.Program;
+
+  if (extra.singleRun && options.filePath) {
+    parseAndGenerateServicesCalls[options.filePath] =
+      parseAndGenerateServicesCalls[options.filePath] || 0;
+    parseAndGenerateServicesCalls[options.filePath] =
+      parseAndGenerateServicesCalls[options.filePath] + 1;
```

Comment:
Can you change this line to the following:
```suggestion
      (parseAndGenerateServicesCalls[options.filePath] || 0) + 1;
```
and then remove 588-9?

Reply:
Yep I agree this increased terseness doesn't sacrifice readbility and will update 👍